rts/idris_gmp.c

```diff
@@ -374,11 +375,37 @@ VAL idris_castFloatBig(VM* vm, VAL f) {
 }
 
 VAL idris_castStrBig(VM* vm, VAL i) {
-    return MKBIGC(vm, GETSTR(i));
+    // TODO: validate that the string is actually an integer
+    return MKBIGC(vm, (char *) GETSTR(i).str_data);
 }
 
 VAL idris_castBigStr(VM* vm, VAL i) {
-    char* str = mpz_get_str(NULL, 10, GETMPZ(GETBIG(vm, i)));
-    return MKSTR(vm, str);
+
+
+    // Compute the appropriate size in bytes for a C string holding the result
+    size_t max_len = mpz_sizeinbase(GETMPZ(GETBIG(vm, i)), 10) + 2;
+    char *temp = (char *)malloc(max_len);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate %lu bytes for bigint as string", (unsigned long)max_len);
+        exit(EXIT_FAILURE);
+    }
+    memset(temp, '\0', max_len);
+    mpz_get_str(temp, 10, GETMPZ(GETBIG(vm, i)));
+
+    unsigned int j;
+    for (j = 0; j < max_len; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j])) && temp[j] != '-') {
```

Comment:
We probably can. Really, my goal is to ensure that all in-memory strings are valid UTF-8, so I want to do this kind of validation for each source of strings.